keylime/allowlistdbproxy.py

```diff
@@ -0,0 +1,128 @@
+import threading
+import time
+
+from sqlalchemy import select
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm.exc import NoResultFound
+
+from keylime import keylime_logging
+from keylime.db.verifier_db import VerifierAllowlist
+
+logger = keylime_logging.init_logging("allowlistdbproxy")
+
+
+class AllowlistEntry:
+    """AllowlistEntry represents a named allowlist in the DB"""
+
+    @staticmethod
+    def from_db(session, allowlist_name):
+        """Create an AllowlistEntry object from an entry in the DB"""
+        try:
+            entry = session.query(VerifierAllowlist).filter_by(name=allowlist_name).one()
+        except NoResultFound:
+            return None
+        except SQLAlchemyError as e:
+            logger.error("SQLAlchemy Error: %s", e)
+            return None
+        # FIXME: pass allowlist.hash or so once it exists
+        return AllowlistEntry(session, allowlist_name, entry.id, entry)
+
+    def __init__(self, session, allowlist_name, checksum, verifier_allowlist, update_timeout=2):
+        """constructor"""
+        self.session = session
+        self.allowlist_name = allowlist_name
+        self.last_update_check = time.time()
+        self.update_timeout = update_timeout
+        self.checksum = checksum
+        self.verifier_allowlist = verifier_allowlist
+
+    def __has_hash_changed(self):
+        """Check whether the unique identifier in the DB has changed"""
+        # FIXME: change to .hash once available
+        statement = select(VerifierAllowlist.id).filter_by(name=self.allowlist_name)
```

Comment:
This line seems to cause issued in the Fedora 34 container where SQLAlchemy package 1.3.22 is likely installed (https://rpmfind.net/linux/RPM/fedora/34/x86_64/p/python3-sqlalchemy-1.3.22-2.fc34.x86_64.html). More recent versions of SQLAlchemy support this statement.

```
2022-11-11 15:34:23.191 - keylime.verifier - ERROR - Operator 'getitem' is not supported on this expression
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/keylime/cloud_verifier_tornado.py", line 1165, in process_agent
    allowlist_entry = allowlist_db_proxy.get_allowlist(allowlist_name)
  File "/usr/local/lib/python3.9/site-packages/keylime/allowlistdbproxy.py", line 117, in get_allowlist
    if not allowlist_entry.check_update():
  File "/usr/local/lib/python3.9/site-packages/keylime/allowlistdbproxy.py", line 54, in check_update
    if self.__has_hash_changed():
  File "/usr/local/lib/python3.9/site-packages/keylime/allowlistdbproxy.py", line 42, in __has_hash_changed
    statement = select(VerifierAllowlist.id).filter_by(name=self.allowlist_name)
```

Reply:
The above is works on RHEL-9 with `SQLAlchemy==1.4.43` and following the tests runs it probably also works with the version in Fedora 35 (1.4.23, https://fedora.pkgs.org/35/fedora-x86_64/python-sqlalchemy-doc-1.4.23-1.fc35.noarch.rpm.html)

Reply:
.where() is supported by older versions